File: leetcode/dp/10. Regular Expression Matching/js/main.js
const isMatch = (s, p) => {
  if (p.length == 0) {
    return s.length == 0;
  }

  let isValid = s.length && (s[0] == p[0] || p[0] == '.');

  if (p.length >= 2 && p[1] == '*') {
    // divide by two
    return isMatch(s, p.substring(2)) || (isValid && isMatch(s.substring(1), p));
  } else {
    if (isValid) {
      return isMatch(s.substring(1), p.substring(1));
    }
    return false;
  }
};

// Time O(m * n)
// Space O(m * n)
const isMatch_II = (s, p) => {
  if (s === null || p === null) {
    return false;
  }

  const dp = Array(s.length + 1)
    .fill(null)
    .map(() => Array(p.length + 1).fill(false));

  dp[0][0] = true;

  for (let i = 1; i < dp[0].length; i++) {
    if (p[i - 1] == '*') {
      dp[0][i] = dp[0][i - 2];
    }
  }

  for (let i = 1; i < dp.length; i++) {
    for (let j = 1; j < dp[0].length; j++) {
      if (p[j - 1] === '.' || s[i - 1] === p[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      } else if (p[j - 1] === '*') {
        // two operations
        dp[i][j] = dp[i][j - 2];

        if (s[i - 1] === p[j - 2] || p[j - 2] === '.') {
          dp[i][j] = dp[i][j] || dp[i - 1][j];
        }
      } else {
        dp[i][j] = false;
      }
    }
  }

  return dp[s.length][p.length];
};

File: leetcode/dp/10. Regular Expression Matching/py/main.py

class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        if not p:
            return len(s) == 0
        
        is_valid = s and (s[0] == p[0] or p[0] == '.')
        
        if len(p) >= 2 and p[1] == '*':
            # if our state is valid s[0] == p[0] we can continue iterate throught s or we finished with *
            return is_valid and self.isMatch(s[1:], p) or self.isMatch(s, p[2:]) 
        else:
            if is_valid:
                return self.isMatch(s[1:], p[1:])
            return False
            
    def isMatch_II(self, s, p):
        memo = {}

        def dfs(s_index, p_index):
          if p_index >= len(p):
            return s_index >= len(s)

          if (s_index, p_index) in memo:
            return memo[s_index, p_index]

          is_valid = s_index < len(s) and (s[s_index] == p[p_index] or p[p_index] == '.')

          if p_index <= len(p) - 2 and p[p_index + 1] == '*':
            if is_valid:
              res = dfs(s_index + 1, p_index) or dfs(s_index, p_index + 2)
              memo[s_index, p_index] = res
              return res
            else:
              res = dfs(s_index, p_index + 2) 
              memo[s_index, p_index] = res
              return res
          else:
            if is_valid:
              res = dfs(s_index + 1, p_index + 1)
              memo[s_index, p_index] = res
              return res

            memo[s_index, p_index] = False  
            return False

        return dfs(0, 0)
